src/Agent.Plugins/PipelineArtifact/FilePathUNCServer.cs

```diff
@@ -0,0 +1,95 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Agent.Sdk;
+using Microsoft.TeamFoundation.Build.WebApi;
+using Microsoft.VisualStudio.Services.WebApi;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.VisualStudio.Services.Content.Common.Tracing;
+
+namespace Agent.Plugins.PipelineArtifact
+{
+    public class FilePathUNCServer
+    {
+        internal async Task UploadAsync(
+            AgentTaskPluginExecutionContext context,
+            Guid projectId,
+            int buildId,
+            string artifactName,
+            string targetPath,
+            string fileSharePath,
+            CancellationToken token)
+        {
+            string artifactPath = Path.Join(fileSharePath, artifactName);
+
+            // create the artifact. at this point, mkdir already succeeded so the path is good.
+            // the artifact should get cleaned up during retention even if the copy fails in the
+            // middle
+            Directory.CreateDirectory(artifactPath);
+
+            // 2) associate the pipeline artifact with an build artifact
+            VssConnection connection = context.VssConnection;
+            BuildServer buildServer = new BuildServer(connection);
+            var propertiesDictionary = new Dictionary<string, string>
+            {
+                { FileShareArtifactUploadEventProperties.ArtifactName, artifactName },
+                { FileShareArtifactUploadEventProperties.ArtifactType, PipelineArtifactConstants.FileShareArtifact },
+                { FileShareArtifactUploadEventProperties.ArtifactLocation, fileSharePath }
+            };
+
+            var artifact = await buildServer.AssociateArtifactAsync(projectId, buildId, artifactName, ArtifactResourceTypes.FilePath, fileSharePath, propertiesDictionary, token);
+            var parallel = context.GetInput(FileShareArtifactUploadEventProperties.Parallel, required: false);
+            var parallelCount = parallel == "true" ? GetParallelCount(context, context.GetInput(FileShareArtifactUploadEventProperties.ParallelCount, required: false)) : 1;
+
+            // To copy all the files in one directory to another directory.
```

Comment:
>in [](start = 37, length = 2)

*from

Reply:
This comment is really copied from build artifact task.